### DISCUSSION TITLE
Spacing between tiles on a watermark

### DISCUSSION THREAD

**Discussion by dekay14** [ImageMagick/ImageMagick#2028] [Help] [answered]
Hello,
I am trying to create a watermark using Text and I am using the following example. As you can see below the label text is pretty long,(actual use case has a much longer text).  The tile needs to be at an angle of 45 degrees reading left to right. I don't seem to understand why the spacing between the tiles increase as the length of the text increases. How can I configure the spacing between the tiles?   Also, would this be the right approch if I have to alternate between fonts on the watermark? Appreciate any help. 

`magick logo.gif \
\( -background none -font Arial -pointsize 18 -fill grey \
label:"This is very long long long long long long long long label." -rotate -45 -write mpr:tile +delete \) \
\( +clone -tile mpr:tile -draw "color 0,0 reset" \) \
-compose over -composite wm_test.gif`

![wm_test](https://user-images.githubusercontent.com/6627708/82127742-c6736400-976a-11ea-8fb5-f5fb40780ce6.gif)

Thank you!

**Comment by fmw42** [comment, marked as answer]
The spacing increases because of the rotation of the text image. You need to tile the text image before rotating.

Try this. Adjust the 30 on the splices to change the spacing. 

```
magick logo: -set option:dims "%[fx:2*w]x%[fx:2*h]" \
\( -background none -font Arial -pointsize 18 -fill grey \
label:"This is very long long long long long long long long label." \
-write mpr:text +delete \) \
\( mpr:text -background none -gravity south -splice 0x30 \
-gravity east -splice 30x0 -write mpr:tile +delete \) \
\( -size "%[dims]" tile:mpr:tile -rotate -45 \) \
-gravity center -compose over -composite wm_test.gif

```

![wm_test](https://user-images.githubusercontent.com/13458218/82128318-118f7600-976f-11ea-91ee-2998a4a2e496.gif)


If you want different fonts, then make two text label: images and append them with space between using -smush and perhaps offset in x using -roll. Then tile that.

**Reply by fmw42** [reply]
It has nothing to do with the output format provided the format supports transparency. So PNG and TIFF work fine. If you need JPG, then you would have to flatten the result against some color, such as white, before saving to JPG, since JPG does not support transparency.

**Reply by dekay14** [reply]
It did work for jpg but like you said; the colors seem off now. That's why I deleted my post later. I pretty new to Imagemagick, I will work with what you have provided and impovise on it. Thanks again for your help!

**Reply by dekay14** [reply]
Thanks @fmw42 ! the following worked for me. Not sure why we would need the use `-roll` though , I included the script with `-roll` as well, it seemed to have shifed the image a little bit.

`magick logo.gif -set option:dims "%[fx:2*w]x%[fx:2*h]" \
\( -background none -font Arial -pointsize 18 -fill black \
label:"This is very long long long long long long long long label that is hard to display with in a image" \
-background none -font Arial -pointsize 18 -fill "rgba(0,0,0,0.5)" \
label:"This is very long long long long long long long long label that is hard to display with in a image" \
-smush 30 \
-write mpr:text +delete \) \
\( mpr:text -background none -gravity south -splice 0x20 \
-gravity east -splice 20x0 -write mpr:tile +delete \) \
\( -size "%[dims]" tile:mpr:tile -rotate -45 \) \
-gravity center -compose over -composite wm_logo_test.gif.jpg`

`magick logo.gif -set option:dims "%[fx:2*w]x%[fx:2*h]" -roll +25+0 \
\( -background none -font Arial -pointsize 18 -fill black \
label:"This is very long long long long long long long long label that is hard to display with in a image" \
-background none -font Arial -pointsize 18 -fill grey \
label:"This is very long long long long long long long long label that is hard to display with in a image" \
-smush 20 \
-write mpr:text +delete \) \
\( mpr:text -background none -gravity south -splice 0x20 \
-gravity east -splice 20x0 -write mpr:tile +delete \) \
\( -size "%[dims]" tile:mpr:tile -rotate -45 \) \
-gravity center -compose over -composite wm_logo_test1.gif.jpg`

![wm_logo_test gif](https://user-images.githubusercontent.com/6627708/82166159-7f6d9780-986c-11ea-8dd1-943f0f056c4b.jpg)
![wm_logo_test1 gif](https://user-images.githubusercontent.com/6627708/82166160-80062e00-986c-11ea-85bd-80c37df5486e.jpg)

**Reply by fmw42** [reply]
Please put your code inside the <> tags so that it is more readable and copyable.

You should roll the second line of text and put some spaces after each line of text.

```
magick logo: -set option:dims "%[fx:2*w]x%[fx:2*h]" \
\( \
    \( -background none -font Arial -pointsize 18 -fill black \
         label:"This is very long long long long long long long long label that is hard to display with in a image.      " \) \
    \( -background none -font Arial -pointsize 18 -fill gray \
         label:"This is very long long long long long long long long label that is hard to display with in a image.      " -roll +50+0 \) \
-smush 20 -write mpr:text +delete \
\) \
\( mpr:text -background none -gravity south -splice 0x20 \
-write mpr:tile +delete \) \
\( -size "%[dims]" tile:mpr:tile -rotate -45 \) \
-gravity center -compose over -composite wm_logo_fred.png

```


![wm_logo_fred](https://user-images.githubusercontent.com/13458218/82168540-d4f97280-9873-11ea-8506-b4e4e3b59231.png)

**Reply by dekay14** [reply]
Thanks you so much!

**Reply by dekay14** [reply]
Hi @fmw42 ,
I am trying to figure out if I can have the text to be continues instead of repeating so that the text is not lost. As you can see some of the text is never displayed unless the resolution of image is large enough. Appreciate your advice.

Thanks!

**Reply by fmw42** [reply]
I do not know what you mean by contuous so that it is not lost? Lost in what sense?  If you have very long text, you can do that provided it covers your image. However, once you rotate the image, there will be empty spaces at the corners, since ImageMagick rotate fills out the image to account for the increased size due to the rotation. Alternately, you would  have to crop the image to the bounds of the text. Can you give me an example of the text you want to use for the image above?

So this is all that ImageMagick can do. It cannot fill to the corners.

```
magick logo: -set option:dims "%wx%h" \
\( -size "%[dims]" -background none -fill gray -gravity center \
caption:"Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum." \) \
-compose over -composite \
-rotate -45 result.png
```


![result](https://user-images.githubusercontent.com/13458218/83723015-d0aeb280-a5f2-11ea-993c-2ff769df198a.png)

**Comment by dekay14** [comment]
The problem is the text size, I need to disply a very long text. The watermarking style does not change (alternating color, text aligned at 45 degree angle etc), the image with watermarking shown above from few weeks ago works just fine. It's the text, you see how each row in the alternating tiles is exactly aligned. This is causing certain part of text to be lost, there might be some text that will never be displayed (especially, text at the begining and end of the text). I am trying to figure out if there is a way for the text to roll over to the next tile, so there is some continuity and the text is not lost. Appreciate your help.

Thanks!

**Reply by fmw42** [reply]
I am still not sure what you want. The alternating text lines in the earlier example are offset. You can adjust the offset using the -roll argument value. In the above example earlier it was -roll +50+0. Increase the +50.  If that does not do what you want, can you mock up an example to show what you desire?

Please note that text can only be wrapped to fill a rectangular shape. The text is written on the image before it is rotated. So when the image is rotated, the text will either not fill the corners as in the last example or get cut off as in the previous examples.

**Reply by dekay14** [reply]
Something similar to the image below (excuse my poor paint skills). Notice the text which is part of the label , the alternating tile does not have the same repeating pattern as the previous row (fo the same color).  If you notice the image from few weeks back, the text in the alternating tiles aligns in the same line. For example the text "This is a very long long log" in the older image; which is repeated on each alternating tile is aligned (if the watermaking was not algined at a 45 degree angle, the alternating text will be in the same column).

![logo_new](https://user-images.githubusercontent.com/6627708/83800234-fa4ff400-a65b-11ea-9d12-ea057df4c9d6.gif)

**Reply by fmw42** [reply]
The only way that I can think to do that is as follows, but it will not let you alternate black and gray. The method is simply to repeat your text over and over.

```
text="This is very long long long long long long long long label that is hard to display with in a image.          This is very long long long long long long long long label that is hard to display with in a image.           This is very long long long long long long long long label that is hard to display with in a image.           This is very long long long long long long long long label that is hard to display with in a image.           This is very long long long long long long long long label that is hard to display with in a image.           This is very long long long long long long long long label that is hard to display with in a image.           This is very long long long long long long long long label that is hard to display with in a image.           This is very long long long long long long long long label that is hard to display with in a image.           This is very long long long long long long long long label that is hard to display with in a image.           This is very long long long long long long long long label that is hard to display with in a image.           This is very long long long long long long long long label that is hard to display with in a image.           This is very long long long long long long long long label that is hard to display with in a image.           This is very long long long long long long long long label that is hard to display with in a image.          This is very long long long long long long long long label that is hard to display with in a image.           This is very long long long long long long long long label that is hard to display with in a image.           This is very long long long long long long long long label that is hard to display with in a image.           This is very long long long long long long long long label that is hard to display with in a image.           This is very long long long long long long long long label that is hard to display with in a image.           This is very long long long long long long long long label that is hard to display with in a image.           This is very long long long long long long long long label that is hard to display with in a image.          This is very long long long long long long long long label that is hard to display with in a image.          "
      
im7 magick logo: -set option:dims "%[fx:2*w]x%[fx:2*h]" \
\( -size "%[dims]" -background none -interline-spacing 24 -pointsize 24 -font Arial -fill black -gravity center caption:"$text" -rotate -45 \) \
-gravity center -compose over -composite wm_logo_fred.png

```


![wm_logo_fred](https://user-images.githubusercontent.com/13458218/83811604-f6799d00-a66e-11ea-8b40-2ef61277fc68.png)


You can also in this case make the white background on the logo transparent and swap the order so that the text does not show on the color parts of the logo.

```
text="This is very long long long long long long long long label that is hard to display with in a image.          This is very long long long long long long long long label that is hard to display with in a image.           This is very long long long long long long long long label that is hard to display with in a image.           This is very long long long long long long long long label that is hard to display with in a image.           This is very long long long long long long long long label that is hard to display with in a image.           This is very long long long long long long long long label that is hard to display with in a image.           This is very long long long long long long long long label that is hard to display with in a image.           This is very long long long long long long long long label that is hard to display with in a image.           This is very long long long long long long long long label that is hard to display with in a image.           This is very long long long long long long long long label that is hard to display with in a image.           This is very long long long long long long long long label that is hard to display with in a image.           This is very long long long long long long long long label that is hard to display with in a image.           This is very long long long long long long long long label that is hard to display with in a image.          This is very long long long long long long long long label that is hard to display with in a image.           This is very long long long long long long long long label that is hard to display with in a image.           This is very long long long long long long long long label that is hard to display with in a image.           This is very long long long long long long long long label that is hard to display with in a image.           This is very long long long long long long long long label that is hard to display with in a image.           This is very long long long long long long long long label that is hard to display with in a image.           This is very long long long long long long long long label that is hard to display with in a image.          This is very long long long long long long long long label that is hard to display with in a image.          "
      
magick logo: -transparent white -set option:dimsA "%wx%h" -set option:dimsB "%[fx:2*w]x%[fx:2*h]" \
\( -size "%[dimsB]" -background white -interline-spacing 24 -pointsize 24 -font Arial -fill black -gravity center caption:"$text" -rotate -45 +repage \) \
+swap -gravity center -compose over -composite -crop "%[dimsA]+0+0" +repage wm_logo_fred2.png

```


![wm_logo_fred2](https://user-images.githubusercontent.com/13458218/83813109-99331b00-a671-11ea-9c54-39af0d754434.png)

**Reply by dekay14** [reply]
Thanks @fmw42

**Reply by dekay14** [reply]
Hi @fmw42 , Could you provide your insight on another issue with respect to watermerking text, is there a co-relation between the font size and the pixels of the image. I have been having a issue with watermarking where the text font seems accurate on some jpg files but the font size is scaled down significantly for others. Also, I noticed that if I read that partiuclar file to an outputsteam, the colors seem to be inverted.  Appreciate your response.

**Reply by fmw42** [reply]
Colors are likely inverted because the JPG is CMYK and likely does not have a proper cmyk profile. For pointsize and pixels, see https://websemantics.uk/tools/font-size-conversion-pixel-point-em-rem-percent/. Note that different browsers and display devices can have different default resolutions. So that could affect the look.

**Reply by dekay14** [reply]
Hi @fmw42,
I am trying a different variation to solve watermarking issue to improve performance. I would like to create a text image (with the watermarking text) with 100% transpernacy and use that image to overlap on the actual image using composite command. 
I am trying to generate the text image using the following command  (two labels and using tiling in order for the text to alternate)

```
magick -size 720x580 xc:none \( -font Arial -pointsize 18 -fill black label:"Lorem ipsum..Lorem ipsum" \ 
-font Arial -pointsize 18 -fill rgba(0,0,0,0.5) label:"Carpe Diam.. Test Text" -roll +50+0 -smush 25 \
 -write mpr:text +delete) \( mpr:text -gravity south -splice 0x20 -write mpr:tile +delete \) \( tile:mpr:tile -rotate -45 \) \ 
-gravity center -compose over -composite wm_test1.jpg
```

What I get is the following:

![wm_test](https://user-images.githubusercontent.com/6627708/107086093-58664580-67ae-11eb-8354-a0503c2f5ec8.jpg)

Can you provide some insight on what I am missing here. I would expect the the label to alternate and the text to repeat but I just see the first label. Appreciate your response.

Thank you!

**Reply by fmw42** [reply]
I am not sure I understand what you want. Could mock up something in GIMP or PS to show me the kind of output you expect?

There is quite a bit strange about your code.  

First, please use the <> keys to format your commands so that the end of line \ are at the end of line and so that you do not present your code in one line.  Second, in IM 7, use magick, not magick convert and not convert.  Third label: keys off -size and -background. So are creating text images from label: using -size 1000x1000, perhaps not what you want. Label: does not recognize xc:none. If you want a transparent background use -background none and specify a -size Wx or -size Hx -size WxH appropriate for the pointsize. Next, I think you may need to put each label: in separate set of parentheses. Label: creates a new image. So have to place each label: image onto your xc:none background image where desired. You either need to combine the two text images and then tile that or tile each separately and then overlay each onto your background image.

Bottom line, I just do not know what you want for the output to be able to correct your code or more likely write it from scratch. I think you are very far from having clean code.

**Reply by dekay14** [reply]
@fmw42 , Appologize for not providing the detailed context. The issue I was working on was to watermark an image which you help me with few months back. But now I am trying to see of I can do the same thing in a performant manner. To achive this, I would like to create a Text image with the watermarking text with 100% transperancy and use this to super impose on the actual jpeg file which needs watermarking. 
Right now am just focused on the first part, which is to create the water marked text image as shown below:

![wm_test](https://user-images.githubusercontent.com/6627708/107102673-10f0b100-67d0-11eb-82d4-50aa2c1ede45.jpg)

here, as you can see the alternating text have different fonts and text. I am trying to use Label to achive this. Also each of these Label needs to be tiled (repeat itself). I tried leveraging what you had suggeseted last time but certainly looks like I am missing or not using the right commands to do this.

**Reply by fmw42** [reply]
You need to create two text images, one for each font and size. Then append them (or smush them if you want spacing). The do the same think as my top post on this page

```
magick logo: -set option:dims "%[fx:2*w]x%[fx:2*h]" \
\( \
\( -background none -font Arial -pointsize 18 -fill black label:"This Is Text1" \) \
\( -background none -font Candice -pointsize 24 -fill black label:"This Is Text2" \) \
+smush 30 -write mpr:text +delete \
\) \
\( mpr:text -background none -gravity south -splice 0x30 -gravity east -splice 30x0 -write mpr:tile +delete \) \
\( -size "%[dims]" tile:mpr:tile -rotate -45 \) \
-gravity center -compose over -composite wm_test.gif

```

![wm_test](https://user-images.githubusercontent.com/13458218/107106167-24574880-67df-11eb-96c2-c84d7c344bd0.gif)

I do not understand your comment about transparent text. If the text is transparent, you will not see it at all.

If you mean do the same on an image that is transparent, then replace logo: with -size WxH xc:none.  HOWEVER, there seems to be a bug and a physical file needs to be created from the transparent image and piped. I suspect it has to do with the -set option.


```
magick -size 640x480 xc:none MIFF:- | magick - -set option:dims "%[fx:2*w]x%[fx:2*h]" \
\( \
\( -background none -font Arial -pointsize 18 -fill black label:"This Is Text1" \) \
\( -background none -font Candice -pointsize 24 -fill black label:"This Is Text2" \) \
+smush 30 -write mpr:text +delete \
\) \
\( mpr:text -background none -gravity south -splice 0x30 -gravity east -splice 30x0 -write mpr:tile +delete \) \
\( -size "%[dims]" tile:mpr:tile -rotate -45 \) \
-gravity center -compose over -composite wm_test2.gif

```

![wm_test2](https://user-images.githubusercontent.com/13458218/107106463-95e3c680-67e0-11eb-986e-51254e79bc7d.gif)

**Reply by dekay14** [reply]
Thank you @fmw42 , that code looks more like my inital post on Friday. I updated the code to use -smush to  alternate the labels as shown below:

![wm_test2](https://user-images.githubusercontent.com/6627708/107161275-72d82480-6950-11eb-9cc8-c4a9e19c1492.gif)

But why don't I get the same result when I try to change the output image to jpg rather than a gif? I just get an image with a black background when I change the output file format to jpg.
Also, why would the text be a bit grainy (doesn't seem clear, not sure if it's just me).

So, I was able to achive what I wanted though using the watermark command
`magick composite -watermark 100%  -gravity South wm_test2.gif  The-Mickey-A-Baseball-Pro.jpg wm_logo2.jpg`

and follwoing is the resulting image:

![wm_logo](https://user-images.githubusercontent.com/6627708/107161380-3b1dac80-6951-11eb-9de5-6e525da41e3e.gif)

But still have few questions about the file format and the grainy-ness of the text. Thanks again Fred! Appreciate all your help.

**Reply by dekay14** [reply]
Also, what method would you prefer for applying watermarking, the above method where I created the watermark image and superimposed it on the actual image to watermark or or what we had suggested earlier:

```
magick logo.gif -set option:dims "%[fx:2*w]x%[fx:2*h]" -roll +25+0  \ 
\( -background none -font Arial -pointsize 18 -fill black  \ 
label:"This is very long long long long long long long long label that is hard to display with in a image"  \
-background none -font Arial -pointsize 18 -fill grey \ 
label:"This is very long long long long long long long long label that is hard to display with in a image"  \ 
-smush 20 \ -write mpr:text +delete \) \ \( mpr:text -background none -gravity south -splice 0x20 \ 
-gravity east -splice 20x0 -write mpr:tile +delete \)  \ 
\( -size "%[dims]" tile:mpr:tile -rotate -45 \) \ 
-gravity center -compose over -composite wm_logo_test1.gif.jpg
```

**Comment by snibgo** [comment]
> I just get an image with a black background when I change the output file format to jpg.

You don't show the command that made that image, so we can only guess. Perhaps you write opaque black text on a transparent black background. JPG cannot record transparency.

> Also, why would the text be a bit grainy (doesn't seem clear, not sure if it's just me).

Please show your command, and any input files.

**Reply by dekay14** [reply]
Hi @snibgo,
Following is the command, am just changing the output image format to a jpg instead of a gif:
```
magick -size 640x480 xc:none MIFF:- | magick - -set option:dims "%[fx:2*w]x%[fx:2*h]" \
\( \
\( -background none -font Arial -pointsize 18 -fill black label:"This Is Text1" \) \
\( -background none -font Candice -pointsize 18 -fill black label:"This Is Text2" \) \
-smush 30 -write mpr:text +delete \
\) \
\( mpr:text -background none -gravity south -splice 0x30 -gravity east -splice 30x0 -write mpr:tile +delete \) \
\( -size "%[dims]" tile:mpr:tile -rotate -45 \) \
-gravity center -compose over -composite wm_test3.jpg
```
resultant image:

![wm_test3](https://user-images.githubusercontent.com/6627708/107164207-dc612e80-6962-11eb-9dbd-6297f421398b.jpg)

regarding the text being grainy, here's the command :

```
magick -size 640x480 xc:none MIFF:- | magick - -set option:dims "%[fx:2*w]x%[fx:2*h]" \
\( \
\( -background none -font Arial -pointsize 18 -fill black label:"This Is Text1" \) \
\( -background none -font Candice -pointsize 18 -fill black label:"This Is Text2" \) \
-smush 30 -write mpr:text +delete \
\) \
\( mpr:text -background none -gravity south -splice 0x30 -gravity east -splice 30x0 -write mpr:tile +delete \) \
\( -size "%[dims]" tile:mpr:tile -rotate -45 \) \
-gravity center -compose over -composite wm_test.gif
```

Thank you!

**Reply by fmw42** [reply]
With regard to the JPG output, you are creating your text on a transparent background and trying to save it as JPG. JPG does not support transparency and the result will be totally black. If you want it to show on JPG, you must composite it onto an opaque background image before saving to JPG. So you either must do it at one time onto some other image, or save using PNG or GIF on transparent background and then later composite it over some other background. If the background is opaque, you can save as JPG; otherwise, save as PNG or GIF or TIF
